libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -901,55 +1024,46 @@ internal class MapRouteLine(
+     * @param point representing the current position of the puck
      */
     fun updateTraveledRouteLine(point: Point) {
-        if (vanishingPointUpdateInhibited) {
+        if (vanishingPointUpdateInhibited ||
+            System.nanoTime() - lastIndexUpdateTimeNano > MAX_ELAPSED_SINCE_INDEX_UPDATE_NANO
+        ) {
             return
         }
 
-        ifNonNull(primaryRoute, primaryRoute?.distance()) { primaryRoute, routeDistance ->
-            val lineString = getLineStringForRoute(primaryRoute)
-
-            if (lineString.coordinates().size < 2) {
-                return
-            }
-
-            val nearestPointOnLineDistance = TurfMisc.nearestPointOnLine(
-                point,
-                lineString.coordinates(),
-                TurfConstants.UNIT_METERS
-            ).getProperty("dist").asDouble
-
-            if (nearestPointOnLineDistance > ROUTE_LINE_UPDATE_MAX_DISTANCE_THRESHOLD_IN_METERS) {
-                return
-            }
-
-            val distanceRemainingFromCache: Float = distanceRemainingCache[primaryRoute]
-                ?: primaryRoute.distance().toFloat()
-            val distanceRemaining = calculatePreciseDistanceTraveledAlongLine(
-                lineString,
-                distanceRemainingFromCache,
-                point
-            )
-            val distanceTraveled = (routeDistance - distanceRemaining)
-            val percentTraveled = abs((distanceTraveled / routeDistance)).toFloat()
-
-            if (percentTraveled > MINIMUM_ROUTE_LINE_OFFSET) {
-                val expression = getExpressionAtOffset(percentTraveled)
-                hideCasingLineAtOffset(percentTraveled)
-                hideRouteLineAtOffset(percentTraveled)
-                decorateRouteLine(expression)
+        ifNonNull(
+            primaryRouteLineGranularDistances,
+            primaryRouteRemainingDistancesIndex
+        ) { granularDistances, index ->
+            val traveledIndex = granularDistances.distancesArray[index]
+            val upcomingPoint = traveledIndex.point
+
+            /**
+             * Take the remaining distance from the upcoming point on the route and extends it
+             * by the exact position of the puck.
+             */
+            val remainingDistance =
+                traveledIndex.distanceRemaining + calculateDistance(upcomingPoint, point)
+
+            /**
+             * Calculate the percentage of the route traveled and update the expression.
+             */
+            if (granularDistances.distance >= remainingDistance) {
+                val offset = (1.0 - remainingDistance / granularDistances.distance)
+                if (offset >= 0) {
+                    val expression = getExpressionAtOffset(offset)
+                    hideCasingLineAtOffset(offset)
+                    hideRouteLineAtOffset(offset)
+                    decorateRouteLine(expression)
+                }
```

Comment:
What about extracting this block of code into a `private` `fun`ction and give it a name based on the comment? This way `updateTraveledRouteLine` will be easier to read and understand and comment will become superfluous so it won't be necessary.